### PR TITLE
feat: 🎸 Title escrow deploy method added

### PR DESCRIPTION
**Pull request by sumitnitsurat**
added a class titlescrow and a method to deploy the contract

- Added a TitleEscrow class and a deploy method in it with 3 param registry, beneficiary and holder.
- Method was provided on etherjs to create a facotry and then deploy it.

 Note-  Can not find the best way to put the method as other places required documents to be supplied. while deploying the contract itself doesn't need a document and its very initial process.

### CODE REVIEW

File: src/titleEscrow.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  wallet: ethers.Wallet;
+
+  constructor({wallet = getWallet()}: {wallet: ethers.Wallet | undefined}) {
+    if (!wallet) {
+      throw new Error("TitleEscrow requires a wallet to be supplied at initialization");
+    }
+    this.wallet = wallet;
+  }
+
+  async deployEscrowContract(
```
Comment (rjchow) [inline src/titleEscrow.ts]: can put this as a static method under the class https://github.com/Open-Attestation/oa-token/blob/a412863b6c94804bb4a681ffabf6392bc828ddc9/src/owner.ts#L80

and take a wallet as a param with getWallet() default

File: src/titleEscrow.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  ) {
+    if (!registryAddress) throw new Error("Please provide the registry address");
+    if (!beneficiaryAddress || !holderAddress)
+      throw new Error("Escrow contract requires beneficiary and holder address");
+
+    const factory = new ethers.ContractFactory(JSON.stringify(TitleEscrowABI), TitleEscrowBytecode, this.wallet);
+    const contract = await factory.deploy(registryAddress, beneficiaryAddress, holderAddress);
+    await contract.deployed();
+
+    return {address: contract.address, hash: contract.deployTransaction.hash};
```
Comment (rjchow) [inline src/titleEscrow.ts]: i think you can just directly return a TitleEscrowOwner class

File: test/escrowContract.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  });
+
+  afterEach("", () => {
+    setWeb3Provider(undefined);
+    setWallet(undefined);
+  });
+
+  describe("Deploy Contract", () => {
+    it("should deploy the contract with passed params", async () => {
+      const ownerInstanceWithHolder1 = await new TitleEscrow({wallet: owner2});
```
Comment (rjchow) [inline test/escrowContract.js]: variable name doesn't match what the class is

File: test/escrowContract.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      const {address} = await ownerInstanceWithHolder1.deployEscrowContract(
+        ERC721Address,
+        owner1Address,
+        holder1Address
+      );
+
+      await ERC721Instance.safeMint(address, SAMPLE_TOKEN_ID);
+      const titleEscrowInstance = new TitleEscrowOwner({address});
+
```
Comment (rjchow) [inline test/escrowContract.js]: can also assert that the token id and title escrow status are correct

**Review by rjchow — COMMENTED** [review]
nice work, just a few small changes :)

File: test/titleEscrowDeploy.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+      expect(await titleEscrowInstance.beneficiary()).to.deep.equal(owner1Address);
+      expect(await ERC721Instance.ownerOf(SAMPLE_TOKEN_ID)).to.deep.equal(titleEscrowInstance.address);
+      expect(await titleEscrowInstance.contractInstance.status()).to.deep.equal(1);
+    });
+
+    it("should throw an error if there is no saved wallet and no wallet was provided in constructor", async () => {
+      const errorDeployTxn = WriteableTitleEscrowOwner.deployEscrowContract({
+        registryAddress: ERC721Address,
+        beneficiaryAddress: owner1Address,
+        holderAddress: holder1Address
```
Comment (rjchow) [inline test/titleEscrowDeploy.js]: Should pass a provider here

File: test/titleEscrowDeploy.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      const errorDeployTxn = WriteableTitleEscrowOwner.deployEscrowContract({
+        registryAddress: ERC721Address,
+        beneficiaryAddress: owner1Address,
+        holderAddress: holder1Address,
+        wallet: owner2
+      });
+      await expect(errorDeployTxn).to.be.rejectedWith(/Deploying contract requires the web3 provider/);
+    });
+
+    it("should throw error if no registry provider", async () => {
```
Comment (rjchow) [inline test/titleEscrowDeploy.js]:
```suggestion
    it("should throw error if no registry address provided", async () => {
```

File: test/titleEscrowDeploy.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+      const errorDeployTxn = WriteableTitleEscrowOwner.deployEscrowContract({
+        registryAddress: undefinedAddress,
```
Comment (rjchow) [inline test/titleEscrowDeploy.js]: Just omit the key?

File: test/titleEscrowDeploy.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        wallet: owner2,
+        web3Provider: provider
+      });
+      await expect(errorDeployTxn).to.be.rejectedWith(/Please provide the registry address/);
+    });
+
+    it("should throw error if no beneficiary or holder address", async () => {
+      const errorDeployTxn = WriteableTitleEscrowOwner.deployEscrowContract({
+        registryAddress: ERC721Address,
+        beneficiaryAddress: owner1Address,
```
Comment (rjchow) [inline test/titleEscrowDeploy.js]: This should be omitted

File: test/titleEscrowDeploy.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        holderAddress: undefinedAddress,
```
Comment (rjchow) [inline test/titleEscrowDeploy.js]: As above

File: test/titleEscrowDeploy.js (hunk 7 above)
Comment (sumitnitsurat) [reply test/titleEscrowDeploy.js]: removed.

File: test/titleEscrowDeploy.js (hunk 8 above)
Comment (sumitnitsurat) [reply test/titleEscrowDeploy.js]: removed

File: test/titleEscrowDeploy.js (hunk 5 above)
Comment (sumitnitsurat) [reply test/titleEscrowDeploy.js]: fixed

**Review by rjchow — APPROVED** [review]

**Pull request merged**